### PR TITLE
[Editor] Add a color picker with predefined colors for highlighting text (bug 1866434)

### PR DESCRIPTION
**Pull request by calixteman**
The doorhanger for highlighting has a basic color picker composed of 5 predefined colors
to set the default color to use.
These colors can be changed thanks to a preference for now but it's something which could
be changed in the Firefox settings in the future.
Each highlight has in its own toolbar a color picker to just change its color.
The different color pickers are so similar (modulo few differences in their styles) that
this patch introduces a new class ColorPicker which provides a color picker component
which could be reused in future editors.
All in all, a large part of this patch is dedicated to color picker itself and its style
and the rest is almost a matter of wiring the component.

### CODE REVIEW

**Review by flodolo — APPROVED** [review]

**Comment by Snuffleupagus** [comment]
Something that would really help (and reduce the time it takes me to review this), given the size/scope of these changes, is a high-level overview of what's being changed and *why* here.
Can you pretty please write a paragraph, or even two, explaining these changes (maybe put it in the *commit message* as well)?

Skimming this, there's a few things that I'm not totally sure of:
 - Do we actually *need* to "generate" the viewer-UI highlightToolbar?
    Why can't we keep the HTML code in the viewer, to (slightly) lessen the amount of changes here?

 - Why are the colors specified in a preference, since that seems to add overall complexity?
   Furthermore, is this intended to be temporary during development/testing (there's no comment about this)?
   Given the new localization-strings, for the colour-names, it doesn't seem (to me) that e.g. the user changing this preference would make a lot of sense?

**Comment by Snuffleupagus** [comment]
*I'm slowly starting to review this, but before looking in detail at the code I've got some observations based on playing with the preview in https://github.com/mozilla/pdf.js/pull/17359#issuecomment-1834007730 above.*

 - Nit: There's a typo in the commit message, `hightlighting text` -> `highlighting text`.

 - Looking at the new editorHighlight viewer-toolbar, the padding feels ever so slightly "off" compared to the existing editor-toolbars:
    - The `padding-top` value seems to be a little bit too large, resulting in what feels like slightly too much empty space at *the top* of the new toolbar.
    - When the left-most color is hovered/selected the focus-ring extends "beyond" the left edge of the "Highlight"-label, 
which again seems a little weird.

   toolbar

 - The new editorHighlight viewer-toolbar probably doesn't handle RTL locales correctly, since the "Highlight"-label is still left-aligned.

   rtl

 - The button *order* on the new "inline" editorHighlight-toolbar seems inconsistent, when compared with the editorStamp-toolbar; note the red outlines in the screen-shot below.
    In my opinion, having the Delete-button "move around" depending on the editor-type doesn't seem great. First of all it just feels inconsistent, and secondly it might lead to users accidentally clicking the Delete-button when moving from one editor-type to another.
    For any button(s) that *all* editors implement, I really think that it should have the same relative position in all editors.
    
    inline

**Comment by caseyr28** [comment]
Hi @Snuffleupagus - UX chiming in here. Thanks for the review and feedback.
1. **editorHighlight viewer-toolbar padding;** Agreed, we will make consistent with the other toolbars.
2. **Left-most icon alignment;** This was designed that intentionally, as otherwise the unselected state would have an indent that would appear off, in my opinion. I am taking some inspiration from the typographic principle of "[Hanging punctuation](https://en.wikipedia.org/wiki/Hanging_punctuation)" - obviously these are not all characters, but optically the color-swatch itself should be the main-aligned element.
3. **The button order on the new "inline" editorHighlight-toolbar;** Agreed, thanks for catching that. To be updated.
4. **handle RTL locales;** @calixteman let me know if we need to add specs for this.

Thanks

File: web/annotation_editor_layer_builder.css [hunk 1]
```diff
@@ -965,7 +997,144 @@
...
+
+      .buttons {
+        .colorPicker {
+          position: relative;
+          width: auto;
+          display: flex;
+          justify-content: center;
+          align-items: center;
+          gap: 4px;
+          padding: 4px;
```
Comment (Snuffleupagus) [inline web/annotation_editor_layer_builder.css:1012]: Here, and possibly below as well, *some* of these values seem to be duplicates of ones defined further up in this file; can we perhaps reduces some unneeded duplication here?

https://github.com/mozilla/pdf.js/blob/319e85449ea6b49bb89a5542def5f88894c812ee/web/annotation_editor_layer_builder.css#L240-L243

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
One thing that I noticed, which might be on purpose but feels a little bit inconsistent, is that after you'd added a highlightEditor its color can *only* be changed with the `editToolbar`-dropdown.
However, trying to change the color from the viewer-toolbar does nothing for a currently selected highlightEditor.

File: src/display/editor/color_picker.js [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    div.setAttribute("aria-multiselectable", false);
+    div.setAttribute("aria-orientation", "vertical");
+    div.setAttribute("data-l10n-id", "pdfjs-editor-colorpicker-dropdown");
+    for (const [name, color] of this.#uiManager.highlightColors) {
+      const button = document.createElement("button");
+      button.tabIndex = "0";
+      button.role = "option";
+      button.setAttribute("data-color", color);
+      button.title = name;
+      button.setAttribute("data-l10n-id", `pdfjs-editor-colorpicker-${name}`);
```
[the remark this answers is not in the document]
Comment (calixteman) [reply src/display/editor/color_picker.js:109]: Yes there is.
If the pref is changed and the color name isn't translated then the value in `title` will be used.

File: web/images/editor-toolbar-colorpicker-arrow.svg
[the remark this answers is not in the document]
Comment (calixteman) [reply web/images/editor-toolbar-colorpicker-arrow.svg:1]: Yep, I just picked up the one the design document.
I raised the question and we'll use the one we've in order to be consistent.

File: web/annotation_editor_layer_builder.css (hunk 1 above)
Comment (calixteman) [reply web/annotation_editor_layer_builder.css:1012]: They aren't at the same level.

**Comment by calixteman** [comment]
> One thing that I noticed, which might be on purpose but feels a little bit inconsistent, is that after you'd added a highlightEditor its color can _only_ be changed with the `editToolbar`-dropdown. However, trying to change the color from the viewer-toolbar does nothing for a currently selected highlightEditor.

Yep it's on purpose: the color in the main toolbar is the default color to use.

File: web/annotation_editor_layer_builder.css [hunk 3]
```diff
@@ -965,7 +997,144 @@
     }
 
     .editToolbar {
+      --editor-toolbar-colorpicker-arrow-image: url(images/toolbarButton-menuArrow.svg);
```
Comment (Snuffleupagus) [inline web/annotation_editor_layer_builder.css:1000]: Please add this file to the following list:
https://github.com/mozilla/pdf.js/blob/319e85449ea6b49bb89a5542def5f88894c812ee/gulpfile.mjs#L1077-L1082

File: web/annotation_editor_layer_builder.css [hunk 4]
```diff
@@ -965,7 +997,144 @@
...
+            justify-content: center;
+            align-items: center;
+            flex-direction: column;
+            gap: 11px;
+            padding-block: 8px;
+            border-radius: 6px;
+            background-color: var(--editor-toolbar-bg-color);
+            border: 1px solid var(--editor-toolbar-border-color);
+            box-shadow: var(--editor-toolbar-shadow);
+            inset-block-start: calc(100% + 3.59px);
```
Comment (Snuffleupagus) [inline web/annotation_editor_layer_builder.css:1045]: Sorry, but I don't really understand why we need to use such a "precise" value here; why can't the pixel-value be an integer?

If there's a good reason for this, then this needs a comment explaining why!

**Review by Snuffleupagus — COMMENTED** [review]
r=me, thank you!

**Pull request merged**